dynamic_programming/abbreviation.cpp

```diff
@@ -0,0 +1,187 @@
+/**
+ * @file
+ * @brief Implementation of Abbrievation
+ * (https://www.hackerrank.com/challenges/abbr/problem)
+ *
+ * @details
+ * Given two strings, `a` and `b`, determine if it's possible to make `a` equal
+ * to `b` You can perform the following operations on the string `a`:
+ * 1. Capitalize zero or more of `a`'s lowercase letters.
+ * 2. Delete all of the remaining lowercase letters in `a`.
+ *
+ * ### Algorithm
+ * The idea is in the problem statement itself: iterate through characters of
+ * string `a` and `b` (for character indexes `i` and `j` respectively):
+ * 1. If `a[i]` and `b[j]` are equal, then move to next position
+ * 2. If `a[i]` is lowercase of `b[j]`, then explore two possibilities:
+ * a. Capitalize `a[i]` or
+ * b. Skip `a[i]`
+ * 3. If the `a[i]` is not uppercase, just discard that character, else return
+ * `false`
+ *
+ * Time Complexity: (O(|a|*|b|)) where `|a|` => length of string `a`
+ * @author [Ashish Daulatabad](https://github.com/AshishYUO)
+ */
+
+#include <cassert>   /// for `assert`
+#include <iostream>  /// for IO operations
+#include <string>    /// `std::string` library
+#include <vector>    /// for `std::vector` STL library
+/**
+ * @namespace dynamic_programming
+ * @brief Dynamic Programming Algorithms
+ */
+namespace dynamic_programming {
+/**
+ * @namespace abbreviation
+ * @brief Implementation of Abbreivation problem
+ */
+namespace abbreviation {
+/**
+ * @brief
+ * Recursive Dynamic Programming function
+ * @details
+ * Returns whether `s` can be converted to `t` with following rules:
+ * a. Capitalize zero or more of a's lowercase letters from string `s`
+ * b. remove all other lowercase letters from string `s`
+ * @param memo memo as parameter to store the result
+ * @param visited visited boolean to check if the result is already computed
+ * @param str given string, which might not be abbreivated
+ * @param result resultant abbreivated string
+ * @param str_idx pointer for string `str`, helpful for transitions
+ * @param result_idx pointer for string `result`, helpful for transitions
+ * @returns `false` if string `str` cannot be converted to `result`
+ * @returns `true` if string `str` can be converted to `result`
+ */
+bool abbreviation_recursion(std::vector<std::vector<bool>> *memo,
+                            std::vector<std::vector<bool>> *visited,
+                            const std::string &str, const std::string &result,
+                            int str_idx = 0, int result_idx = 0) {
```

Comment:
Consider using `uint64_t`, `uint32_t`, `uint16_t` or even `uint8_t` for non-negative values. For negative values, use `int64_t` or their appropriate size. Check other parts of the code too ([reference](https://en.cppreference.com/w/cpp/types/integer)). 🙂